### PR TITLE
Exact 2Q state preparation

### PR DESCRIPTION
**Pull request by ecpeterson**
A long while back, Oscar Perdomo provided us with some formulas to perform exact state-to-state transformations between two-qubit systems. This PR implements his techniques.

TODO:
* [x] The case where the source is unentangled and the target is entangled is handled by reversing the state prep instruction, compiling that, and then reversing the output. _Very_ confusingly, this seems to result in a loss of precision. Addressing this is probably a requirement before merging.
* [x] Write to Oscar to announce the inclusion of these methods.
* [x] Change `#+#:ignore` to whatever the informative attic feature is.

### CODE REVIEW

**Comment by jmbr** [comment]
Are the notes available somewhere?

File: src/compilers/state-prep.lisp [hunk 1]
```diff
@@ -229,6 +232,109 @@
...
+(define-compiler state-prep-2Q-compiler
+    ((instr (_ _ q1 q0)
+            :where (typep instr 'state-prep-application)))
+  "Exact, optimal compiler for STATE-PREP-APPLICATION instances that target a pair of qubits."
+  (flet ((orthogonal-vector (v)
+           (make-row-major-matrix 2 1 (list (- (conjugate (magicl:ref v 1 0))) (conjugate (magicl:ref v 0 0)))))
+         (normalize-vector (v)
+           (let ((norm (sqrt (+ (expt (abs (magicl:ref v 0 0)) 2)
+                                (expt (abs (magicl:ref v 1 0)) 2)))))
+             (print norm) (terpri)
```
Comment (appleby) [inline src/compilers/state-prep.lisp]: 👀

File: src/compilers/state-prep.lisp [hunk 2]
```diff
@@ -145,11 +146,236 @@
...
+                                                        0 0 -1 0
+                                                        0 0  0 1))))
+          (setf matrix (m* m matrix))
+          (setf v (nondestructively-apply-matrix-to-vector matrix vector)))))
     (list matrix v)))
 
+(defun state-prep-within-local-equivalence-class (source-wf target-wf)
+  "Produces a circuit that carries a two-qubit wavefunction SOURCE-WF to a two-qubit target wavefunction TARGET-WF under the hypothesis that SOURCE-WF and TARGET-WF have the same Segre obstruction.
 
+Returns a pair (LIST C0 C1) of 2x2 matrices with (C0 (x) C1) SOURCE-WF = TARGET-WF."
```
Comment (stylewarning) [inline src/compilers/state-prep.lisp]: can you state why you return a pair and not values?

File: src/compilers/state-prep.lisp [hunk 3]
```diff
@@ -145,11 +146,236 @@
...
+                    (d (normalize-vector (m* A b-vector)))
+                    (z (normalize-vector (m* B x-vector)))
+                    (u (normalize-vector (m* B y-vector)))
+                    (theta1 (- pi/2 (phase (+ (sqrt (first lambdas)) (* #C(0 1) (sqrt (second lambdas)))))))
+                    (theta2 (- pi/2 (phase (+ (sqrt (first deltas)) (* #C(0 1) (sqrt (second deltas)))))))
+                    (U1 (make-row-major-matrix 2 2 (list (magicl:ref c 0 0) (magicl:ref d 0 0)
+                                                         (magicl:ref c 1 0) (magicl:ref d 1 0))))
+                    (U2 (make-row-major-matrix 2 2 (list (magicl:ref z 0 0) (magicl:ref u 0 0)
+                                                         (magicl:ref z 1 0) (magicl:ref u 1 0))))
+                    (V1 (make-row-major-matrix 2 2 (mapcar #'conjugate
```
Comment (stylewarning) [inline src/compilers/state-prep.lisp]: this mapcar is extremely @ecpeterson

File: src/compilers/state-prep.lisp (hunk 2 above)
Comment (ecpeterson) [reply src/compilers/state-prep.lisp]: Yeah: the two operators don't make sense in isolation. Elsewhere in the code, I've used returning `values` to mean "you might need the other return values, but perhaps not" and returning a `list` to mean "you definitely need all the values, so you'd better be ready to destructure them". It helps me typecheck receiving code that I write.

Comment (stylewarning) [reply src/compilers/state-prep.lisp]: can you comment this in the code as a nice reminder?

is it too much to `defstruct foo (:type list)` ?

Comment (ecpeterson) [reply src/compilers/state-prep.lisp]: I pushed a version that uses `adt:defdata`.

**Pull request merged**